provider/specs/client.md

```diff
@@ -0,0 +1,82 @@
+An implementation of a OFREP client provider should start with an initialization of the provider.
+
+The `initialize()` function should follow those steps:
+1. Make a GET request to the `/ofrep/v1/configuration` endpoint to retrieve the configurations return by the flag management system and store them in memory to be available for all the function of the provider. *(See [Annexe 1](#annexe-1) for the description of the endpoint response)*
+2. Make a POST request to the `/ofrep/v1/evaluate/flags` endpoint with the evaluation context in the body.  
+   - If the endpoint return an error, the `initialize()` function must error and exit.  
+   - If the endpoint is in success, we should store in a local cache all the flags evaluation result returned by the API in a local cache. We should also store the `ETag` header in the provider to be able to send it back later.
+3. If polling is enabled, the function should start the polling loop *(See [polling section](#polling))*.
+
+## Evaluation
+The evaluation should not do any remote API calls.
+
+When calling an evaluation function the provider should check if the associated type is supported, by checking the key `capabilities.flagEvaluation.unsupportedTypes` from the configuration endpoint.
+- If the type is unsupported, we should exit early and directly return an error.
+- If the type is supported, we should retrieve the flag from the local cache of the flags evaluation.
+  - If you are not able to find the flag, you should return an FlagNotFound error.
+  - If the remote evaluation for this flag has return an error, you should map the error in provider and return the associated error.
+  - If the value retrieve from the cache has a different type than the one expected you should return a TypeMismatch error.
+  - If the cached evaluation is in success you should return the evaluation response.
+
+## Polling
+The polling system will make a POST request periodically tp the `/ofrep/v1/evaluate/flags` endpoint to check if there is a change in the flags evaluation to be able to store it.
+
+If we have a `ETag` available we should always add the header `If-None-Match` with the `ETag` value.
+- If the cache is still up-to-date we will receive a `304` telling us that the nothing has changed on the flag management system side.
+- If the cache is outdated we will receive a `200` with the new values of all the flags. In that situation we should:
+  1. Replace the actual local cache of flags evaluations.
+  2. Store the new `ETag` value for the future call.
+
+
+## Annexe 1
+**`/ofrep/v1/configuration` description:**
+The endpoint will return a list of configuration for the provider:
+- `name`: Name of the flag management system, it should be used in the `metadata.name` name (ex: `OFREP Web Provider ${metadata.name}`).
+- `capabilities`: List of capabilities of the flag management system and their associated configuration. *Check the [capabilities section](#capabilities) for the details of each capability.*
+
```

Comment:
What's the expected behavior if the call to `/ofrep/v1/configuration` fails?

Reply:
Should fallback to the defaults and I think we should define the default behaviour in this spec

Reply:
Hum good question, we were wondering with @lukas-reining and I think we should put the provider in error and retry later (probably with an exponential backoff).

WDYT?

Reply:
If we want to configure it, I will be more in favour of having an opt-out rather than an opt-in.

The reasoning is that this will encourage vendors to provide the information on how it works but also to avoid using the default values, the information of this endpoint is here to help to work better with the flag management system, not using them by default will be a bit weird from my point of view.

Reply:
This will affect initial page load times, and the config API doesn't provide anything (at least not yet) that couldn't be defined explicitly by the OFREP constructor.